### ISSUE TITLE
Task: Fix ESlint Errors (tracking issue)

### ISSUE THREAD

**Issue by scinos** [Automattic/wp-calypso#56330]
### Details

Current list of errors (updated to commit 6a130575c5759ba153f1310e9de36cb4629a19a4)

| Rule                                                 | Errors | Disabled | Autofix| Issue
| ---------------------------------------------------- | ------:| --------:| ------ | --
| [wpcalypso/jsx-classname-namespace][1]               | 273    | 308      |        | 
| [jsx-a11y/anchor-is-valid][6]                        | 24     | 21       |        | 
| [jsx-a11y/alt-text][7]                               | 23     | 5        |        | 
| [jsx-a11y/click-events-have-key-events][9]           | 19     | 31       |        | 
| [jsx-a11y/no-static-element-interactions][11]        | 18     | 22       |        | 
| [jsdoc/no-undefined-types][3]                        | 17     | 5        |        | #56607
| [no-restricted-imports][14]                          | 15     | 15       |        | 
| [no-console][15]                                     | 14     | 100      |        | 
| [no-shadow][16]                                      | 10     | 9        |        | 
| [no-unused-vars][18]                                 | 7      | 18       |        | 
| [eqeqeq][20]                                         | 6      | 3        | ✅      | 
| [react/no-danger][21]                                | 5      | 25       |        | 
| [no-case-declarations][22]                           | 5      | 19       |        | 
| [import/order][24]                                   | 11     | 4        | ✅      | 
| [wpcalypso/jsx-gridicon-size][25]                    | 4      | 19       |        | 
| [import/named][26]                                   | 4      | 3        |        | #56981 
| [wpcalypso/redux-no-bound-selectors][27]             | 3      | 17       |        | 
| [wpcalypso/i18n-ellipsis][28]                        | 3      | 0        | ✅      | 
| prettier/prettier                                    | 3      | 0        | ✅      | 
| [jsx-a11y/heading-has-content][30]                   | 3      | 5        |        | 
| [jest/no-export][31]                                 | 3      | 2        |        | 
| [@typescript-eslint/no-use-before-define][32]        | 3      | 21       |        | 
| [react/jsx-no-target-blank][33]                      | 2      | 4        | ✅      | 
| [no-undef][34]                                       | 2      | 24       |        | 
| [no-nested-ternary][35]                              | 2      | 8        |        | 
| [no-global-assign][36]                               | 2      | 2        |        | 
| [jsx-a11y/no-noninteractive-element-interactions][37]| 2      | 7        |        | 
| [jsx-a11y/no-autofocus][38]                          | 2      | 41       |        | 
| [jsx-a11y/iframe-has-title][39]                      | 2      | 1        |        | 
| [wpcalyspo/jsx-classname-namespace][44]              | 1      | 0        |        | 
| [wpcalypso/i18n-mismatched-placeholders][45]         | 1      | 1        |        | 
| [react/no-did-update-set-state][46]                  | 1      | 4        |        | 
| [jsx-a11y/role-supports-aria-props][47]              | 1      | 0        |        | 
| [jsx-a11y/html-has-lang][48]                         | 1      | 0        |        | 
| [import/no-nodejs-modules][50]                       | 1      | 36       |        | 

## Fixed errors

These errors have been already fixed. The list is here for documentation purposes:

| Rule                                                 | Errors | Disabled | Autofix|Issue|
| ---------------------------------------------------- | ------:| --------:| ------|--|
| [jsdoc/require-param][2]                             | 76     | 4        | ✅|#56577|
| [jsdoc/no-undefined-types][3]                        | 65     | 5        | |#56607|
| [jsdoc/check-param-names][4]                         | 50     | 0        | |#56604|
| [react/no-string-refs][5]                            | 32     | 3        |        | #56719
| [@typescript-eslint/ban-types][8]                    | 21     | 7        | |#56612|
| [jsdoc/require-param-description][10]                | 19     | 0        | |#56577|
| [jest/expect-expect][12]                             | 17     | 17       | |#56698|
| [jsdoc/check-tag-names][13]                          | 16     | 15       | |#56605|
| [jest/no-conditional-expect][17]                     | 9      | 4        | |#56721|
| [jsdoc/newline-after-description][19]                | 6      | 0        | ✅|#56606|
| [jsdoc/require-returns-description][23]              | 5      | 0        | |#56577|
| [jsdoc/check-values][41]                             | 2      | 1        | |#56609|
| [jsdoc/check-alignment][42]                          | 2      | 0        | ✅|#56580|
| [jest/no-identical-title][43]                        | 2      | 1        | |#56698|
| [jest/valid-title][49]                               | 1      | 3        | |#56698|
| [jsx-a11y/accessible-emoji][40]                      | 2      | 0        | ||

[1]: https://github.com/Automattic/wp-calypso/blob/trunk/packages/eslint-plugin-wpcalypso/docs/rules/jsx-classname-namespace.md
[2]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/require-param.md
[3]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/no-undefined-types.md
[4]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/check-param-names.md
[5]: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-string-refs.md
[6]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md
[7]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/alt-text.md
[8]: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md
[9]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/click-events-have-key-events.md
[10]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/require-param-description.md
[11]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-static-element-interactions.md
[12]: https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/expect-expect.md
[13]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/check-tag-names.md
[14]: https://eslint.org/docs/rules/no-restricted-imports
[15]: https://eslint.org/docs/rules/no-console
[16]: https://eslint.org/docs/rules/no-shadow
[17]: https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/no-conditional-expect.md
[18]: https://eslint.org/docs/rules/no-unused-vars
[19]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/newline-after-description.md
[20]: https://eslint.org/docs/rules/eqeqeq
[21]: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-danger.md
[22]: https://eslint.org/docs/rules/no-case-declarations
[23]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/require-returns-description.md
[24]: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/order.md
[25]: https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/jsx-gridicon-size.md
[26]: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/named.md
[27]: https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/redux-no-bound-selectors.md
[28]: https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/i18n-ellipsis.md
[30]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/heading-has-content.md
[31]: https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/no-export.md
[32]: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-use-before-define.md
[33]: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-target-blank.md
[34]: https://eslint.org/docs/rules/no-undef
[35]: https://eslint.org/docs/rules/no-nested-ternary
[36]: https://eslint.org/docs/rules/no-global-assign
[37]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-noninteractive-element-interactions.md
[38]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-autofocus.md
[39]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/iframe-has-title.md
[40]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/accessible-emoji.md
[41]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/check-values.md
[42]: https://github.com/gajus/eslint-plugin-jsdoc/blob/master/.README/rules/check-alignment.md
[43]: https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/no-identical-title.md
[44]: https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/jsx-classname-namespace.md
[45]: https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/i18n-mismatched-placeholders.md
[46]: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-did-update-set-state.md
[47]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/role-supports-aria-props.md
[48]: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/html-has-lang.md
[49]: https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/valid-title.md
[50]: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-nodejs-modules.md

**Comment by scinos** [comment]
I've been looking at the lint rules related to JSDoc params, and I think there is room for improvement.

Given that we don't generate external docs from the source code, and I assume that most devs use VSCode (or an editor with very similar features regarding code intellisense), I based this analysis on the usefulness of JSDoc `@param` for VSCode in two scenarios: code intellisense when typing and when mouse hover the function.

**With no `@param` documented**

|On hover|On type|
|-|-|
|![no-param-on-hover](https://user-images.githubusercontent.com/975703/134886436-f23ca4c3-e87f-4a13-8338-72e6f90e6209.png)| ![no-param-on-type](https://user-images.githubusercontent.com/975703/134886448-9d1dd90e-cf90-4733-a65f-cdf71acb3f9f.png)|

Note how VSCode correctly infers the parameters name and types them as `any`.

**With just `@param <name>`**
|On hover|On type|
|-|-|
|![param-on-hover](https://user-images.githubusercontent.com/975703/134886754-a4c0e79e-871f-41c4-bad0-44a4aa040d99.png)|![param-on-type](https://user-images.githubusercontent.com/975703/134886786-e31bef14-b824-4944-b11e-ef5c74853fdb.png)|

Intellisense when typing is exactly the same. On hover it adds the parameter names after the function description, but that info was already visible in the function signature.

**With `@param <type> <name>`**
|On hover|On type|
|-|-|
|![type-on-hover](https://user-images.githubusercontent.com/975703/134887016-1fa7b320-e956-4269-bae8-2b2e923b7ac4.png)|![type-on-type](https://user-images.githubusercontent.com/975703/134887006-db54dfff-4b3d-463c-a9b2-c96ecc6621fb.png)|

Now it shows the types in the function signature in both cases, but it doesn't in param list after the function description

**With `@param <type> <name> <description>`**
|On hover|On type|
|-|-|
|![desc-on-hover](https://user-images.githubusercontent.com/975703/134887232-132fd392-54ce-49c8-8182-e8b26c401567.png)|![desc-on-type](https://user-images.githubusercontent.com/975703/134887240-5ac38af1-f743-4fda-bf4f-3b62d9929c42.png)|

Shows the description in both cases.

---

Given the results above, I think adding just `@param <name>` **doesn't provide any value**. As such, I think we should disable `jsdoc/require-param` but keep `jsdoc/check-param-names` (don't force adding `@param`, but if a `@param` is present, ensure it matches the actual param name).

Enforcing param description is also counter productive. In many cases the param information is available in the function description, can be easily inferred by the name or there is simply no relevant description. Enforcing it will likely cause lots of useless `@param <string> id - The id` The same applies to `jsdoc/require-returns-description`

Now only `@param <type> <name>` is left. We can enforce it by enabling `jsdoc/require-param-type`. While arguably it adds some value, the reality is we'll have lots errors (348 errors) if we enable it. Fixing them is not a trivial task.

Type checks won't be enforced (at least my VSCode config ignores any type discrepancy). Given that we already have infrastructure in place to transpile TypeScript in most packages (example: `client/` is already a TypeScript + JavaScript project), if we think types are important enough to be mandatory, maybe we should convert those files to TypeScript.

In short, I propose:
- Disable `jsdoc/require-param`
- Disable `jsdoc/require-returns-description`
- Disable `jsdoc/require-param-description`
- Keep `jsdoc/no-undefined-types`
- Keep `jsdoc/check-param-names`

**Comment by jsnajdr** [comment]
For me personally, I only find useful the rules that say "I was able to detect there's something obviously wrong about your jsdoc comment", ones that enforce some consistency, like newline between description and params or `{boolean}` and `{String}` type names over `{bool}` and `{string}`.

On the other hand, rules that say "you need to fill in this required information" mostly lead to jsdoc comments that are repetitive, like:
```js
/**
 * Function that converts dogs to cats
 *
 * @param {Dog} dog the dog
 * @returns {Cat} cat the cat
 */
function dogToCat( dog ) {
  /* ... */
  return cat;
}
```
where the jsdoc comment doesn't really say anything new.

This anti-pattern is very common in our Redux code where we are forced to document action creators as:
```js
/**
 * @returns {Action} an action object to be dispatched
 */
```
or reducers as:
```js
/**
 * @param state The original state
 * @param action An action object
 * @returns Updated state
 */
```

I would be better if the linter let the author decide what's worth documenting and let them omit the parts that are obvious.

**Comment by noahtallen** [comment]
I don't have much to add, but I agree with your logic and the proposal to change these rules.

If one wants great type analysis for a JS function, I think we should be encouraging them to re-write in TS rather than jsdoc. I guess it's even possible that jsdoc gives a false sense of security that types are checked by something before deploy.

**Comment by IanVS** [comment]
@scinos I'm excited to see such great use of eslint-nibble, and I hope it's been helpful.  Let me know if you have any suggestions or requests for it.
